src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningConditionMapper.java

```diff
@@ -0,0 +1,426 @@
+                if (fieldName.startsWith("#")) {
+                    // if it's a field placeholder, remove it -- we'll generated a new one for translated expression.
+                    // (we don't need to worry about the field placeholder also being used in the filter expression,
+                    // since the filter expression can contain only non-key fields.)
+                    fieldName = removeFieldPlaceholder(fieldName);
+                }
+
+                if (primaryKey.getHashKey().equals(fieldName)) {
+                    checkArgument(hkValue == null, "Multiple hash key conditions");
+                    ComparatorContext comparator = keyConditionPart.comparator();
+                    checkArgument(comparator != null && comparator.getText().equals("="),
+                        "Hash key condition must have operator '='");
+                    String valuePlaceholder = keyConditionPart.literal(0).getText();
+                    hkValue = getValueAndRemovePlaceholderIfNeeded(valuePlaceholder);
+                } else if (primaryKey.getRangeKey().isPresent() && primaryKey.getRangeKey().get().equals(fieldName)) {
+                    checkArgument(rkCondition.isEmpty(), "Multiple range key conditions");
+                    ComparisonOperator op = getRangeKeyComparisonOp(keyConditionPart);
+                    List<AttributeValue> values = keyConditionPart.literal().stream()
+                        .map(LiteralContext::getText)
+                        .map(this::getValueAndRemovePlaceholderIfNeeded)
+                        .collect(Collectors.toList());
+                    rkCondition = Optional.of(new KeyFieldCondition(op, values));
+                } else {
+                    throw new IllegalArgumentException("Key condition expression contains non-key field: " + fieldName);
+                }
+            }
+            return visitChildren(keyConditionPart);
+        }
+
+        private String removeFieldPlaceholder(String placeholder) {
+            String literal = null;
+            if (fieldPlaceholders != null) {
+                literal = fieldPlaceholders.remove(placeholder);
+            }
+            checkArgument(literal != null, "Referenced attribute name does not exist: %s", placeholder);
+            return literal;
+        }
+
+        private AttributeValue getValueAndRemovePlaceholderIfNeeded(String placeholder) {
+            AttributeValue value = null;
+            if (valuePlaceholders != null) {
+                // keep the value placeholder if it's also in the filter expression; otherwise remove it
+                value = isPlaceholderInFilterExpression(placeholder)
+                    ? valuePlaceholders.get(placeholder)
+                    : valuePlaceholders.remove(placeholder);
+            }
+            checkArgument(value != null, "Referenced attribute value does not exist: %s", placeholder);
+            return value;
+        }
+
```

Comment:
There was a bug both here and also in random partitioning, where a query request has a key condition expression and filter expression, and there's a value placeholder that appears in both expressions. Before I was always removing the value placeholder here and that's wrong if it also appears in the filter expression. For checking whether whether a placeholder appears in the filter expression is based on patterns/matchers because it seems overkill to define the filter expression grammar just for this purpose? Though we actually did start defining one already, though it's incomplete.

Reply:
Actually now I'm thinking I'll define the full grammar in ANTLR and use its parser instead. It seems bad to compile patterns so often, though I don't have a good idea of what the cost would be.

Reply:
OK then I realized I don't need to define the full parser rule for filter/condition expressions and can just use lexer rules. So that's how we get the value placeholders from the filter expression now.